src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs

```diff
@@ -493,6 +495,24 @@ public static string GetMSBuildFilePath(PackageSpec project, RestoreRequest requ
 
                 props.AddRange(GenerateGroupsWithConditions(buildPropsGroup, isMultiTargeting, frameworkConditions));
 
+                // Create an empty PropertyGroup for package properties
+                var packagePathsPropertyGroup = MSBuildRestoreItemGroup.Create("PropertyGroup", Enumerable.Empty<XElement>(), 1000, isMultiTargeting ? frameworkConditions : Enumerable.Empty<string>());
+
+                var projectGraph = targetGraph.Graphs.FirstOrDefault();
+
+                // Autogenerate a property for any packages that have tools or have opted into having a property generated
+                var packagePathProperties = packagesWithTools.Union(projectGraph.Item.Data.Dependencies.Where(i => i.GeneratePathProperty).Select(i => i.Name)).Distinct(StringComparer.OrdinalIgnoreCase)
+                    .Select(i => sortedPackages.Cast<KeyValuePair<LockFileTargetLibrary, Lazy<LocalPackageSourceInfo>>?>().FirstOrDefault(x => x.Value.Key.Name.Equals(i, StringComparison.OrdinalIgnoreCase)))
+                    .Where(i => i != null && i.Value.Value.Exists())
+                    .Select(i => GeneratePackagePathProperty(i.Value.Key.Name, i.Value.Value.GetAbsolutePath(i.Value.Value.Value.Package.ExpandedPath)));
```

Comment:
Can you please simply this logic? these complex linq queries are never easy to read and understand. besides, i also dont understand this `i.Value.Value.Value.Package.ExpandedPath`

Reply:
I agree, the Linq statements in this file really tested my understanding.  I'll break them up a little.